### PR TITLE
#84 create validator bulk send emails with test async

### PR DESCRIPTION
**Pull request by alesyalis**

### CODE REVIEW

File: Recrutify/Recrutify.Services/Validators/CandidatePrimarySkillValidator.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+using Recrutify.Services.DTOs;
+
+namespace Recrutify.Services.Validators
+{
+    public class CandidatePrimarySkillValidator : AbstractValidator<CandidatePrimarySkillDTO>
+    {
+        public CandidatePrimarySkillValidator()
+        {
+            RuleFor(c => c.Id)
+                .NotNull()
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Validators/CandidatePrimarySkillValidator.cs:14]: how Guid can be null?

Comment (alesyalis) [reply Recrutify/Recrutify.Services/Validators/CandidatePrimarySkillValidator.cs]: no, done

File: Recrutify/Recrutify.Services/DTOs/CandidateCreateDTO.cs [hunk 2]
```diff
@@ -2,6 +2,6 @@
 {
     public class CandidateCreateDTO : BaseCandidateDTO
     {
-        public CandidatePrimarySkillDTO PrimarySkill { get; set; }
+        public CreateCandidatePrimarySkillDTO PrimarySkill { get; set; }
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/DTOs/CandidateCreateDTO.cs:5]: Guid PrimarySkillId

File: Recrutify/Recrutify.Services/DTOs/CreateCandidatePrimarySkillDTO.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+﻿using System;
+
+namespace Recrutify.Services.DTOs
+{
+    public class CreateCandidatePrimarySkillDTO
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/DTOs/CreateCandidatePrimarySkillDTO.cs:5]: remove

File: Recrutify/Recrutify.Host/Configuration/Profiles/CandidateProfile.cs [hunk 4]
```diff
@@ -18,7 +18,11 @@ public CandidateProfile()
                 .ForMember(dest => dest.IsAssigned, conf => conf.MapFrom(src => false));
             CreateMap<ProjectResult, ScheduleCandidateProjectResultDTO>().ReverseMap()
                 .ForMember(dest => dest.PrimarySkill, conf => conf.MapFrom(src => new CandidatePrimarySkillDTO { Id = src.PrimarySkill.Id, Name = src.PrimarySkill.Name }));
-            CreateMap<CandidatePrimarySkill, CandidatePrimarySkillDTO>().ReverseMap();
+            CreateMap<CandidatePrimarySkill, CandidatePrimarySkillDTO>()
+                .ForMember(dest => dest.Name, conf => conf.MapFrom(src => new CandidatePrimarySkill { Id = src.Id, Name = src.Name }));
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Host/Configuration/Profiles/CandidateProfile.cs:22]: ignore

File: Recrutify/Recrutify.Host/Configuration/OData/CandidatesConfiguration.cs [hunk 5]
```diff
@@ -10,7 +10,7 @@ public void Apply(ODataModelBuilder builder, ApiVersion apiVersion, string route
         {
             builder.EntitySet<CandidateDTO>("Candidates");
             builder.ComplexType<ProjectResultDTO>();
-            builder.ComplexType<CandidatePrimarySkillDTO>();
+            builder.ComplexType<CreateCandidatePrimarySkillDTO>();
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Host/Configuration/OData/CandidatesConfiguration.cs:13]: are you sure?

File: Recrutify/Recrutify.Host/Configuration/Profiles/CandidateProfile.cs [hunk 6]
```diff
@@ -18,7 +18,11 @@ public CandidateProfile()
...
+            CreateMap<CreateCandidatePrimarySkillDTO, CandidatePrimarySkill>()
+                .ForMember(dest => dest.Name, opt => opt.Ignore());
+
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Host/Configuration/Profiles/CandidateProfile.cs:25]: warninga

File: Recrutify/Recrutify.Services/Services/CandidateService.cs [hunk 7]
```diff
@@ -91,7 +93,10 @@ public async Task<CandidateDTO> CreateAsync(CandidateCreateDTO candidateCreateDT
         {
             var candidate = _mapper.Map<Candidate>(candidateCreateDTO);
             var currentCandidate = await _candidateRepository.GetByEmailAsync(candidate.Email);
+            var currentPrimarySkill = await _primarySkillService.GetAsync(candidateCreateDTO.PrimarySkill.Id);
+            var primarySkillName = currentPrimarySkill.Name;
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/CandidateService.cs:97]: remove lines 97-99

File: Recrutify/Recrutify.Services/Validators/CreateCandidateValidator.cs [hunk 8]
```diff
@@ -35,14 +35,19 @@ public CreateCandidateValidator()
             RuleForEach(c => c.Contacts)
                 .NotNull()
                 .NotEmpty();
-            RuleFor(c => c.Location)
+            RuleFor(c => c.Location.City)
+                .NotNull()
+                .NotEmpty();
+            RuleFor(c => c.Location.Country)
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Validators/CreateCandidateValidator.cs:41]: don't you want validate length of city and country names

File: Recrutify/Recrutify.Services/Services/CandidateService.cs [hunk 9]
```diff
@@ -91,7 +93,10 @@ public async Task<CandidateDTO> CreateAsync(CandidateCreateDTO candidateCreateDT
         {
             var candidate = _mapper.Map<Candidate>(candidateCreateDTO);
             var currentCandidate = await _candidateRepository.GetByEmailAsync(candidate.Email);
+            var currentPrimarySkill = await _primarySkillService.GetAsync(candidateCreateDTO.PrimarySkill.Id);
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/CandidateService.cs:96]: throw ValidationException if currentPrimarySkill is null

File: Recrutify/Recrutify.Host/Configuration/Profiles/CandidateProfile.cs (hunk 4 above)
Comment (elizavetapavliv) [reply Recrutify/Recrutify.Host/Configuration/Profiles/CandidateProfile.cs:22]: remove .ForMember(dest => dest.Name, conf => conf.MapFrom(src => new CandidatePrimarySkill { Id = src.Id, Name = src.Name })); it at all

File: Recrutify/Recrutify.Services/Services/PrimarySkillService.cs [hunk 10]
```diff
@@ -23,5 +24,11 @@ public async Task<List<PrimarySkillDTO>> GetAllAsync()
             var skills = await _primarySkillRepository.GetAllAsync();
             return _mapper.Map<List<PrimarySkillDTO>>(skills);
         }
+
+        public async Task<PrimarySkillDTO> GetAsync(Guid id)
+        {
+            var skills = await _primarySkillRepository.GetAsync(id);
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/PrimarySkillService.cs:30]: why skill**s**?

File: Recrutify/Recrutify.Services/DTOs/CandidateCreateDTO.cs (hunk 2 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/DTOs/CandidateCreateDTO.cs]: done

File: Recrutify/Recrutify.Services/DTOs/CreateCandidatePrimarySkillDTO.cs (hunk 3 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/DTOs/CreateCandidatePrimarySkillDTO.cs]: done

File: Recrutify/Recrutify.Host/Configuration/Profiles/CandidateProfile.cs (hunk 6 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Host/Configuration/Profiles/CandidateProfile.cs]: done

File: Recrutify/Recrutify.Services/Services/CandidateService.cs (hunk 7 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Services/CandidateService.cs]: done

File: Recrutify/Recrutify.Services/Validators/CreateCandidateValidator.cs (hunk 8 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Validators/CreateCandidateValidator.cs:42]: done

File: Recrutify/Recrutify.Services/Services/CandidateService.cs (hunk 9 above)
Comment (alesyalis) [reply Recrutify/Recrutify.Services/Services/CandidateService.cs]: done

**Review by elizavetapavliv — APPROVED** [review]

**Pull request merged**